tao_compiler/mlir/xla/ral/context/cuda_impl_sparse.cc

```diff
@@ -0,0 +1,301 @@
+      cache;
+};
+
+#if defined(PLATFORM_ALIBABA) and defined(ENABLE_BLADE_GEMM)
+template <typename T>
+bladnn::Dtype toBlaDNNDtype() {
+  if (std::is_same<T, Eigen::half>::value) {
+    return bladnn::Dtype::kF16;
+  }
+  if (std::is_same<T, float>::value) {
+    return bladnn::Dtype::kF32;
+  }
+  if (std::is_same<T, double>::value) {
+    return bladnn::Dtype::kF64;
+  }
+  if (std::is_same<T, uint16_t>::value) {
+    return bladnn::Dtype::kU16;
+  }
+  return bladnn::Dtype::kUnknown;
+}
+#endif
+
+template <typename InT, typename OutT, typename MetaT>
+void ral_spgemm(ExecutionContext* ctx, void* stream_handle,
+                MemRefType<InT, 2> A, MemRefType<InT, 2> B,
+                MemRefType<OutT, 2> C, int cd_a, int cd_b) {
+  InT* dA = A.data;
+  InT* dB = B.data;
+  OutT* dC = C.data;
+
+  int64_t num_A_rows = A.sizes[0];
+  int64_t num_A_cols = A.sizes[1];
+  int64_t num_B_rows = B.sizes[0];
+  int64_t num_B_cols = B.sizes[1];
+
+  bool tp_a = cd_a == 1 ? false : true;
+  bool tp_b = cd_b == 0 ? false : true;
+
+  int64_t m = tp_b ? num_B_rows : num_B_cols;
+  int64_t n = tp_a ? num_A_cols : num_A_rows;
+  int64_t k = tp_b ? num_B_cols : num_B_rows;
+
+  float alpha = 1.0f;
+  float beta = 0.0f;
+
+  std::string unique_name =
+      "tao_ral.gpu.sparse_gemm_" + tao::ral::TaoTypeNameHelper<InT>::Invoke();
+  auto state = ctx->getOrCreateResource<SparseGemmState<InT, MetaT>>(
+      unique_name, []() { return new SparseGemmState<InT, MetaT>; });
+  opaque_t compressed_weight, compressed_meta;
+  {
+    GpuSparseGemmKey key{m, n, k, 1, tp_a, tp_b, dB};
+    std::lock_guard<std::mutex> l(state->mu);
+
+    auto& cache = state->cache;
+    auto it = cache.find((intptr_t)dB);
+    if (it == cache.end()) {
+      std::shared_ptr<SparseCompressedWeight<InT, MetaT>> compressed_weight_ptr(
+          new SparseCompressedWeight<InT, MetaT>(ctx, stream_handle, key));
+      it = cache.insert(std::make_pair((intptr_t)dB, compressed_weight_ptr))
+               .first;
+    }
+    compressed_weight = it->second->compressed_weight();
+    compressed_meta = it->second->compressed_meta();
+  }
+
+#if defined(PLATFORM_ALIBABA) and defined(ENABLE_BLADE_GEMM)
+  {
+    auto gpu_driver = ctx->getDriver<GPUDriver>(GPUDriver::name());
+    auto stream =
+        static_cast<se::Stream*>(gpu_driver->asSEStream(ctx, stream_handle));
+    void* s = stream->implementation()->GpuStreamHack();
+    bladnn::Context bladnn_ctx{s};
+    bladnn::Dtype in_dtype = toBlaDNNDtype<InT>();
+    bladnn::Dtype out_dtype = toBlaDNNDtype<OutT>();
+    bladnn::Dtype meta_dtype = toBlaDNNDtype<MetaT>();
+    bool ret = bladnn::spgemm(
+        &bladnn_ctx, in_dtype, !tp_b, compressed_weight, tp_b ? m : k / 2,
+        tp_b ? k / 2 : m, in_dtype, !tp_a, dA, tp_a ? k : n, tp_a ? n : k,
+        out_dtype, dC, m, n, meta_dtype, compressed_meta, k / 32, m * 2);
+
+    if (ret) {
+      return;
+    }
+  }
+#endif
+  ctx->signalError(Context::FAILURE, "unsupport sparse gemm.");
```

Comment:
if ret == 0, it means that bladnn::spgemm run failed, and the result is wrong. In this situation, we should directly exit because we do not have a fallback op.

Reply:
Done, add #elif